nexus/src/db/collection_attach.rs

```diff
@@ -0,0 +1,1494 @@
+        QuerySqlType<ResourceTable<ResourceType, Self>>: Send,
+        // Allows calling ".filter()" on the boxed collection table.
+        BoxedQuery<CollectionTable<ResourceType, Self>>: FilterBy<Eq<CollectionPrimaryKey<ResourceType, Self>, Self::Id>>
+            + FilterBy<IsNull<Self::CollectionTimeDeletedColumn>>,
+        // Allows calling ".filter()" on the boxed resource table.
+        BoxedQuery<ResourceTable<ResourceType, Self>>: FilterBy<Eq<ResourcePrimaryKey<ResourceType, Self>, Self::Id>>
+            + FilterBy<Eq<Self::ResourceCollectionIdColumn, Self::Id>>
+            + FilterBy<IsNull<Self::ResourceCollectionIdColumn>>
+            + FilterBy<IsNull<Self::ResourceTimeDeletedColumn>>,
+
+        // Allows calling "update.into_boxed()"
+        UpdateStatement<
+            ResourceTable<ResourceType, Self>,
+            ResourceTableDefaultWhereClause<ResourceType, Self>,
+            V,
+        >: BoxableUpdateStatement<ResourceTable<ResourceType, Self>, V>,
+        // Allows calling
+        // ".filter(resource_table().primary_key().eq(resource_id)" on the
+        // boxed update statement.
+        BoxedUpdateStatement<'static, Pg, ResourceTable<ResourceType, Self>, V>:
+            FilterBy<Eq<ResourcePrimaryKey<ResourceType, Self>, Self::Id>>,
+
+        // Allows using "id" in expressions (e.g. ".eq(...)") with...
+        Self::Id: AsExpression<
+                // ... The Collection table's PK
+                SerializedCollectionPrimaryKey<ResourceType, Self>,
+            > + AsExpression<
+                // ... The Resource table's PK
+                SerializedResourcePrimaryKey<ResourceType, Self>,
+            > + AsExpression<
+                // ... The Resource table's FK to the Collection table
+                SerializedResourceForeignKey<ResourceType, Self>,
+            >,
+        ExprSqlType<CollectionPrimaryKey<ResourceType, Self>>: SingleValue,
+        ExprSqlType<ResourcePrimaryKey<ResourceType, Self>>: SingleValue,
+        ExprSqlType<Self::ResourceCollectionIdColumn>: SingleValue,
+
+        // Necessary to actually select the resource in the output type.
+        ResourceType: Selectable<Pg>,
+    {
+        let collection_table =
+            || <CollectionTable<ResourceType, Self> as HasTable>::table();
+        let resource_table =
+            || <ResourceTable<ResourceType, Self> as HasTable>::table();
+
+        // Create new queries to determine if the collection and resources
+        // already exist.
+        let collection_exists_query = Box::new(
+            collection_table()
+                .into_boxed()
+                .filter(collection_table().primary_key().eq(collection_id))
+                .filter(Self::CollectionTimeDeletedColumn::default().is_null()),
+        );
+        let resource_exists_query = Box::new(
+            resource_table()
+                .into_boxed()
+                .filter(resource_table().primary_key().eq(resource_id))
```

Comment:
As mentioned in [the linked PrimaryKey](https://docs.diesel.rs/master/diesel/prelude/trait.Table.html#associatedtype.PrimaryKey) documentation, I'm concerned about the case of

> If the table has a composite primary key, this will be a tuple.

where `ResourceIdColumn` is not a tuple. Though, looking at that bound, I can squint and convince myself that if `primary_key()` returns a tuple then compilation would fail :)

Reply:
Yeah, I think the `primary_key()` and `Self::Id` types would need to match for this to work... since the `CollectionIdColumn` must implement `Column`, I think this CTE as currently implemented would not allow for composite PKs.

... but as you said, that would at least be a compile-time error, rather than a runtime error.